source/extensions/filters/http/common/compressor/compressor.cc

```diff
@@ -0,0 +1,308 @@
+
+Http::FilterHeadersStatus CompressorFilter::decodeHeaders(Http::HeaderMap& headers, bool) {
+  if (config_->runtime().snapshot().featureEnabled(config_->featureName(), 100) &&
+      isAcceptEncodingAllowed(headers)) {
+    compressor_ = config_->makeCompressor();
+    skip_compression_ = false;
+    if (config_->removeAcceptEncodingHeader()) {
+      headers.removeAcceptEncoding();
+    }
+  } else {
+    config_->stats().not_compressed_.inc();
+  }
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterHeadersStatus CompressorFilter::encodeHeaders(Http::HeaderMap& headers,
+                                                          bool end_stream) {
+  if (!end_stream && !skip_compression_ && isMinimumContentLength(headers) &&
+      isContentTypeAllowed(headers) && !hasCacheControlNoTransform(headers) &&
+      isEtagAllowed(headers) && isTransferEncodingAllowed(headers) && !headers.ContentEncoding()) {
+    sanitizeEtagHeader(headers);
+    insertVaryHeader(headers);
+    headers.removeContentLength();
+    headers.insertContentEncoding().value(config_->contentEncoding());
+    config_->stats().compressed_.inc();
+  } else if (!skip_compression_) {
+    skip_compression_ = true;
+    config_->stats().not_compressed_.inc();
+  }
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterDataStatus CompressorFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  if (!skip_compression_) {
+    config_->stats().total_uncompressed_bytes_.add(data.length());
+    compressor_->compress(data, end_stream ? Compressor::State::Finish : Compressor::State::Flush);
+    config_->stats().total_compressed_bytes_.add(data.length());
+  }
+  return Http::FilterDataStatus::Continue;
+}
+
+Http::FilterTrailersStatus CompressorFilter::encodeTrailers(Http::HeaderMap&) {
+  if (!skip_compression_) {
+    Buffer::OwnedImpl empty_buffer;
+    compressor_->compress(empty_buffer, Compressor::State::Finish);
+    config_->stats().total_compressed_bytes_.add(empty_buffer.length());
+    encoder_callbacks_->addEncodedData(empty_buffer, true);
+  }
+  return Http::FilterTrailersStatus::Continue;
+}
+
+bool CompressorFilter::hasCacheControlNoTransform(Http::HeaderMap& headers) const {
+  const Http::HeaderEntry* cache_control = headers.CacheControl();
+  if (cache_control) {
+    return StringUtil::caseFindToken(cache_control->value().getStringView(), ",",
+                                     Http::Headers::get().CacheControlValues.NoTransform);
+  }
+
+  return false;
+}
+
+bool CompressorFilter::isAcceptEncodingAllowed(const Http::HeaderMap& headers) const {
+  const Http::HeaderEntry* accept_encoding = headers.AcceptEncoding();
+  if (!accept_encoding) {
+    config_->stats().no_accept_header_.inc();
+    return false;
+  }
+
+  using encPair = std::pair<absl::string_view, float>; // pair of {encoding, q_value}
+  std::vector<encPair> pairs;
+
+  std::vector<std::string> allowed_compressors = config_->registeredCompressors();
```

Comment:
probably this should be a map too?

I realize there won't be a ton of these. Probably just 2 for a long time, but it seems like a natural map.